### PR TITLE
Add “medium freighter” category

### PR DESCRIPTION
**Pull request by Nescio0**
Warships are subdivided into light, medium, and heavy; the criterion seems to be the empty mass (i.e. hull plus outfit capacity).
```
                    mass
light  warship:     <500
medium warship: 500-1000
heavy  warship:    >1000
```

Freighter categories are currently assigned rather arbitrarily. This pull request changes that to harmonize freighter categories to those of warships.
The Hauler, Hauler II, and Hauler III all fit within the medium category (cf. their warship counterparts Nest, Roost, and Skein), as do a few others.

### CODE REVIEW

File: data/ships.txt [hunk 1]
```diff
@@ -317,12 +317,12 @@ ship "Bastion"
 		"hull" 4200
 		"required crew" 32
 		"bunks" 47
-		"mass" 580
+		"mass" 540
```
Comment (jafdy) [inline data/ships.txt]: Why are you changing an existing ship?

File: data/ships.txt [hunk 2]
```diff
@@ -1923,7 +1923,7 @@ ship "Modified Argosy"
 	plural "Modified Argosies"
 	sprite "ship/modified argosy"
 	attributes
-		category "Light Warship"
+		category "Medium Warship"
```
Comment (jafdy) [inline data/ships.txt]: Why are you changing a ship to an existing category, if it was wanted in this category then it would have been put in this category in the first place.

**Comment by jafdy** [comment]
#3249

File: data/ships.txt (hunk 2 above)
Comment (tehhowch) [reply data/ships.txt]: I would agree that the Mod. Argosy is hardly a Medium Warship.

Comment (Bladewood) [reply data/ships.txt]: Notably, modified argosy is on par with a gunboat or a raven in terms of health, but it carries the armament of a firebird or a corvette.

Comment (Nescio0) [reply data/ships.txt]: Because I believe the modified argosy was only assigned the "light warship" category because the normal argosy was in the "light freighter" category because there used to be no "medium freighter" category.

Comparing the modified argosy with other human space warships, it's actually heavier than several medium warships, therefore I believe this change is justified.

File: data/ships.txt (hunk 1 above)
Comment (Nescio0) [reply data/ships.txt]: Because the Bastion used to be heavier than e.g. the Vanguard warship, and I think it ought to remain a medium warship, therefore I believe slightly removing some unnecessary mass is justified here. (With this change the Bastion's free outfit space becomes 14/450, comparable to other medium warships.)

File: data/ships.txt (hunk 2 above)
Comment (Bladewood) [reply data/ships.txt]: Mass *is not* the only thing that determines a ship's category. Otherwise, the Summer Leaf would be an interceptor, or Met Par Tek 53 and Model 32 would be medium warships, or some other ridiculous nonsense would occur. A Marauder Bounder or a Marauder Fury or even a stock Hawk would become a light warship.

File: data/ships.txt (hunk 1 above)
Comment (Bladewood) [reply data/ships.txt]: Uh, mass is independent of outfit space...

**Comment by Nescio0** [comment]
A comparison of human space medium warships:
```
                     shields, hull,    mass + outfit, weapons, engines, guns + turrets, bays
  Aerie            :    5700, 1900,  570 = 130 + 390,     150,      95,      5 = 2 + 3,    2
  Bastion          :    6700, 4200, 1050 = 580 + 470,     180,     120,      7 = 4 + 3,    0
  Corvette         :    6100, 1200,  570 = 150 + 420,     150,     100,      6 = 4 + 2,    0
  Firebird         :    6400, 2800,  690 = 290 + 400,     160,     100,      6 = 4 + 2,    0
  Manta            :    5900, 1500,  520 = 270 + 350,     140,     100,      6 = 6 + 0,    0
 *Modified Argosy* :    4800, 1900,  590 = 250 + 340,     140,      80,      6 = 4 + 2,    0 *
  Mule             :    5400, 4400,  770 = 320 + 450,     210,     110,      7 = 2 + 4,    1
  Nest             :    2500, 3700,  650 = 250 + 400,     140,      80,      6 = 2 + 4,    2
  Osprey           :    7200, 1600,  720 = 270 + 450,     180,     130,      6 = 4 + 2,    0
  Roost            :    2900, 5200,  810 = 360 + 450,     140,      80,      6 = 2 + 4,    4
  Skein            :    6700, 3300,  970 = 470 + 500,     140,      80,      6 = 2 + 4,    6
  Splinter         :    5200, 1700,  650 = 250 + 400,     150,     100,      5 = 2 + 3,    0
```

File: data/ships.txt (hunk 2 above)
Comment (Nescio0) [reply data/ships.txt]: If not mass, then what? If outfits, weapons, engines, number of guns and turrets, the Modified Argosy clearly fits within the medium warship category, see below.

Comment (jafdy) [reply data/ships.txt]: In case you hadn't noticed, there is not set metric on how to define a ship.

Comment (Nescio0) [reply data/ships.txt]: No rules set in stone, perhaps, but there appears to be a pattern, at least for human space ships. I suggest we follow the warship category pattern and apply this to the freighter categories.

**Comment by tehhowch** [comment]
There are some spreadsheets available that rank the various ships/outfits, made by @Amazinite : https://drive.google.com/drive/folders/0B635z_nU19WfQllrM2V2dWpFSFk

I will say that based on your table it does seem sensible to promote the Modified Argosy from Light to Medium Warship. I think it's a bit tough to argue for the creation of a new ship category that has only 5 ships total though, and I've no opinion on the changes to the Bastion.

**Comment by Nescio0** [comment]
> I  will say that based on your table it does seem sensible to promote the  Modified Argosy from Light to Medium Warship. I think it's a bit tough  to argue for the creation of a new ship category that has only 5 ships  total though, and I've no opinion on the changes to the Bastion.
The "Heavy Freighter" category is even smaller; however, I don't think the number of ships should really matter. Why I think this category is necessary, is because "Medium Freighters" are actually statistically quite similar to "Medium Warships", and quite different from both light and heavy freighters.

**Comment by Nescio0** [comment]
Number of freighters in each category with this diff:
```
            Light, Medium,  Heavy,  Total
Human    :      3,      4,      2,      9       
Hai      :      1,      1,      0,      2
Coalition:      3,      4,      2,      9
Wanderer :      1,      0,      1,      2
Korath   :      0,      0,      1,      1
Total    :      8,      9,      6,     23
```

**Comment by Amazinite** [comment]
@ discussion on changing the Bastion and Mod Argosy.
This PR is for the purpose of creating a medium freighter category. The title and description make this clear. What this is not is a PR designed to standardize ship categories based off of certain criteria. Given this, I suggest you remove the Bastion and Mod Argosy changes as they are outside of the scope of this PR as is, and they make the PR less likely to be merged because of this.

As for my opinion on the actual changes:
The change to the Bastion seems very unnecessary to me. Yes, the Bastion is indeed heavier than the Vanguard. It is also heavier than the Behemoth, Falcon, Protector, and Leviathan. The change from 580 mass to 540 mass doesn't actually change any of that, except for making its mass equal that of the Behemoth. If looking at mass + outfit space, then the Vanguard is actually 10 tons heavier than the Bastion, and the ships below the Bastion are the Behemoth and Falcon (which would have the same mass as the Bastion of 1,050) and the Skein (another medium warship who would be at 970 tons). So whether the change in mass is needed or not, your argumentation for it just doesn't work out.
<hr/>

Ignoring the fact that I disagree with the mass change, the following is a graph showing the relationship between mass and outfit space for all ships in the game.
screenshot_271

While more mass does generally mean more outfit space, this is only because more mass is also tied to being a generally bigger ship, which is what is causing the higher outfit space, i.e. it isn't the mass. There's also a wide range of outfit space at any selection of mass. E.g. almost directly between 500 and 1,000 we can see 7 ships with very similar masses but a wide range of outfit spaces that go from what looks like 300 to almost 1000.

The design philosophy of MZ (@endless-sky) for ships also has to do with the size of a ship. He's stated in the past that he will create the ship model, size it, and then based off of the size he will create the stats. In addition to the size, the context of the ship matters as well. (In this case, the context is that the Bastion is a Southbound ship.) In short, the Bastion got lighter, not smaller, so the outfit space shouldn't change. That's not even mentioning the fact that you didn't even check the outfit space of the Bastion's two variants, the Bastion (Laser) and Bastion (Heavy), as both max out their outfit space in vanilla, meaning they'd have -20 outfit space from this PR, and any change to their outfits might warrant balance testing of Free World campaign missions to make sure the Bastion is still pulling its weight in Free Worlds fleets for battles against enemies.
<hr/>

**tl;dr, I disagree with the Bastion changes and believe that they should be reverted in this PR.**
<hr/>

Onto the Modified Argosy. Going back to the previously mentioned importance of context in balancing ships, the Modified Argosy is intended to be heavier set for its category. It's just an Argosy with a bunch of armor plates attached to it and some redesigned innards done by pirates.

Onto the question of stats. Here is a table of the top three light warships and bottom three medium warships in terms of hit points (hull + shields)

Ship | Category | Hit Points
------|---------|----------
Corvette | Medium War | 7,300
Gunboat | Light War | 7,200
Splinter | Medium War | 6,900
Argosy | Light War | 6,700
Nest | Medium War | 6,200
Raven | Light War | 6,100

When considering context, this makes sense to me:
* The Gunboat is top-notch military technology and therefore is expected to have higher hit points for its tier.
* The Nest is a Hauler I converted into a carrier, and therefore will have lower hit points for its category
* The Splinter is built by the Syndicate, who's ships are generally fragile (see Vanguard and Protector being the weakest heavy warships).
* The Corvette is a lightly-built ship designed for speed, and therefore it sacrifices strength.

The Argosy could have a place among the medium warships category, but then we'd have to argue over the placement of the Nest, Gunboat, and possibly the Splinter.
<hr/>
Looking at outfit, weapon, and engine space, here is how the Argosy stacks up to a few medium warships by the metric of outfit space:

Ship | Outfit | Weapon | Engine
-----|-------|-------|------|
Firebird | 400 | 160 | 100
Splinter | 400 | 150 | 100
Nest | 400 | 140 | 80
Aerie | 390 | 150 | 95
Manta | 350 | 140 | 100
Argosy | 340 | 140 | 80

In terms of these metrics, it's clear to see that should the Modified Argosy be a medium warship, it would be the weakest among them. The weapon space is similar to the Nest and Manta, but the Manta has higher outfit and engine space while the Nest has much higher outfit space. None of these medium warships fall lower than the Modified Argosy, and those that tie in certain categories exceed the Argosy in others.
<hr/>

So context in mind, the Modified Argosy can just barely be argued to be a medium warship in terms of hit points and should it be a medium warship would be the weakest in terms of outfit, weapon, and engine space.

Quickly making mention of turret and gun space, the Modified Argosy really only beats out (assuming guns > turrets) the Manta medium warship, which has 6 guns. The Splinter is comparable at 2 guns and 3 turrets, and the Firebird, Osprey, and Corvette match the Modified Argosy. Any other medium warships are better gunned.
<hr/>

**tl;dr The Modified Argosy is currently a powerful light warship, but changing its category only makes it the weakest medium warship, so a change in category is, in my eyes, unnecessary, but not incorrect.**

**Comment by endless-sky** [comment]
There are currently 90 warships in the game, and only 29 freighters. That's why there needed to be 3 different size classes for warships. Given that there's only a third as many freighters, there's no good reason to split them up any further.

**Comment by KiLEdEnNis** [comment]
eventually I think we should add a Medium freighter category, and maybe even split the Transports up in Light, Medium & Heavy
but for now and probably the next couple of years it shouldn't be needed (although I personally prefer Light/Medium/Heavy tabs for all 3)

**Comment by Nescio0** [comment]
* Reverted Bastion tweaks, as it was clear that would not be accepted.
* Spun off  Modified Argosy class change into a new diff.


Let's focus the discussion on the actual proposal (adding a medium freighter category).
Arguments against:
* It does not add much
* There are fewer freighters than warships

Arguments in favour:
* It does not break anything
* Organizing freighters parallel to warships is less arbitrary than the current situation
* More freighters might be added
* Other categories are even smaller

If the number of freighters really is the only objection, then you might as well argue that light and heavy freighters ought to be merged into a single freighter category (cf. transports).

**Comment by jafdy** [comment]
"does not brake anything" is not an argument in favor. 
When more freighters are added then an extra category can be added, once the category is added it can never be removed.
"other categories are even smaller" is a useless argument because those categories can't be merged in any coherent way. 

You appear to have confused "not against" with "in favor".

**Comment by Nescio0** [comment]
> You appear to have confused "not against" with "in favor".

True, however, the arguments "against" are actually also "not in favour".

> once the category is added it can never be removed.

Nonsense. Simply remove a line of code from the `Ship.cpp` file and use `grep` to find all occurences or `sed` to automatically replace them. It could be done in a minute or so, if desired.

**Comment by tehhowch** [comment]
Yes deleting on code is fine. However there is the issue that removing category support will result in a breaking change that can invalidate save files. Endless Sky has 100% backwards compatibility by design to prevent this, so adding support for an attribute (or equivalent item that is saved to or read from the save) means it must be supported forever.

**Comment by KiLEdEnNis** [comment]
not directly relevant
"once the category is added it can never be removed"
not really good to have "Transport" then, when in the future it may be split in Light/Medium/Heavy
w/ Freighters ya just add a category, here ya'd add 3 & remove 1 if that can even be done safely

**Comment by endless-sky** [comment]
I see no good reason to do this.
